From 74a0d751fdd6b2c59255b754deffee335038c07a Mon Sep 17 00:00:00 2001
From: Carlos Alberto Lopez Perez <clopez@igalia.com>
Date: Wed, 27 Feb 2013 03:03:59 +0100
Subject: [PATCH] Fix endianness issues #1073 #1021 #1022

 * Ensure that pkh.len is swapped along with pkh.caplen
---
 src/airbase-ng.c     |  4 +++-
 src/aircrack-ng.c    |  8 ++++++--
 src/airdecap-ng.c    |  4 +++-
 src/aireplay-ng.c    | 16 ++++++++++++----
 src/airodump-ng.c    |  4 +++-
 src/airtun-ng.c      |  4 +++-
 src/ivstools.c       |  4 +++-
 src/packetforge-ng.c |  4 +++-
 src/tkiptun-ng.c     |  4 +++-
 9 files changed, 39 insertions(+), 13 deletions(-)

diff --git a/src/airbase-ng.c b/src/airbase-ng.c
index 92dc7a7..813b2be 100644
--- a/src/airbase-ng.c
+++ b/src/airbase-ng.c
@@ -4767,8 +4767,10 @@ usage:
                 continue;
             }
 
-            if( dev.pfh_in.magic == TCPDUMP_CIGAM )
+            if( dev.pfh_in.magic == TCPDUMP_CIGAM ) {
                 SWAP32( pkh.caplen );
+                SWAP32( pkh.len );
+            }
 
             n = caplen = pkh.caplen;
 
diff --git a/src/aircrack-ng.c b/src/aircrack-ng.c
index 4ab8486..14a270a 100644
--- a/src/aircrack-ng.c
+++ b/src/aircrack-ng.c
@@ -910,8 +910,10 @@ void read_thread( void *arg )
 			if( close_aircrack )
 				break;
 
-			if( pfh.magic == TCPDUMP_CIGAM )
+			if( pfh.magic == TCPDUMP_CIGAM ) {
 				SWAP32( pkh.caplen );
+				SWAP32( pkh.len );
+			}
 
 			if( pkh.caplen <= 0 || pkh.caplen > 65535 )
 			{
@@ -1837,8 +1839,10 @@ void check_thread( void *arg )
 			while( ! atomic_read( &rb, fd, sizeof( pkh ), &pkh ) )
 				goto read_fail;
 
-			if( pfh.magic == TCPDUMP_CIGAM )
+			if( pfh.magic == TCPDUMP_CIGAM ) {
 				SWAP32( pkh.caplen );
+				SWAP32( pkh.len );
+			}
 
 			if( pkh.caplen <= 0 || pkh.caplen > 65535 )
 			{
diff --git a/src/airdecap-ng.c b/src/airdecap-ng.c
index 953d8d7..96251ff 100644
--- a/src/airdecap-ng.c
+++ b/src/airdecap-ng.c
@@ -618,8 +618,10 @@ usage:
         if( fread( &pkh, 1, n, f_in ) != (size_t) n )
             break;
 
-        if( magic == TCPDUMP_CIGAM )
+        if( magic == TCPDUMP_CIGAM ) {
             SWAP32( pkh.caplen );
+            SWAP32( pkh.len );
+        }
 
         n = pkh.caplen;
 
diff --git a/src/aireplay-ng.c b/src/aireplay-ng.c
index 7eb379e..2be9563 100644
--- a/src/aireplay-ng.c
+++ b/src/aireplay-ng.c
@@ -895,8 +895,10 @@ int capture_ask_packet( int *caplen, int just_grab )
                 return( 1 );
             }
 
-            if( dev.pfh_in.magic == TCPDUMP_CIGAM )
+            if( dev.pfh_in.magic == TCPDUMP_CIGAM ) {
                 SWAP32( pkh.caplen );
+                SWAP32( pkh.len );
+            }
 
             tv.tv_sec  = pkh.tv_sec;
             tv.tv_usec = pkh.tv_usec;
@@ -2526,8 +2528,10 @@ int do_attack_arp_resend( void )
                 continue;
             }
 
-            if( dev.pfh_in.magic == TCPDUMP_CIGAM )
+            if( dev.pfh_in.magic == TCPDUMP_CIGAM ) {
                 SWAP32( pkh.caplen );
+                SWAP32( pkh.len );
+            }
 
             tv.tv_sec  = pkh.tv_sec;
             tv.tv_usec = pkh.tv_usec;
@@ -2937,8 +2941,10 @@ int do_attack_caffe_latte( void )
                 continue;
             }
 
-            if( dev.pfh_in.magic == TCPDUMP_CIGAM )
+            if( dev.pfh_in.magic == TCPDUMP_CIGAM ) {
                 SWAP32( pkh.caplen );
+                SWAP32( pkh.len );
+            }
 
             tv.tv_sec  = pkh.tv_sec;
             tv.tv_usec = pkh.tv_usec;
@@ -3363,8 +3369,10 @@ int do_attack_migmode( void )
                 continue;
             }
 
-            if( dev.pfh_in.magic == TCPDUMP_CIGAM )
+            if( dev.pfh_in.magic == TCPDUMP_CIGAM ) {
                 SWAP32( pkh.caplen );
+                SWAP32( pkh.len );
+            }
 
             tv.tv_sec  = pkh.tv_sec;
             tv.tv_usec = pkh.tv_usec;
diff --git a/src/airodump-ng.c b/src/airodump-ng.c
index 467974b..fd533ac 100644
--- a/src/airodump-ng.c
+++ b/src/airodump-ng.c
@@ -6267,8 +6267,10 @@ usage:
                 continue;
             }
 
-            if( G.pfh_in.magic == TCPDUMP_CIGAM )
+            if( G.pfh_in.magic == TCPDUMP_CIGAM ) {
                 SWAP32( pkh.caplen );
+                SWAP32( pkh.len );
+            }
 
             n = caplen = pkh.caplen;
 
diff --git a/src/airtun-ng.c b/src/airtun-ng.c
index 2476c8b..855c6de 100644
--- a/src/airtun-ng.c
+++ b/src/airtun-ng.c
@@ -1592,8 +1592,10 @@ usage:
                 continue;
             }
 
-            if( dev.pfh_in.magic == TCPDUMP_CIGAM )
+            if( dev.pfh_in.magic == TCPDUMP_CIGAM ) {
                 SWAP32( pkh.caplen );
+                SWAP32( pkh.len );
+            }
 
             n = caplen = pkh.caplen;
 
diff --git a/src/ivstools.c b/src/ivstools.c
index 766413f..d3f3f42 100644
--- a/src/ivstools.c
+++ b/src/ivstools.c
@@ -882,8 +882,10 @@ int main( int argc, char *argv[] )
         if( fread( &pkh, 1, n, f_in ) != (size_t) n )
             break;
 
-        if( pfh.magic == TCPDUMP_CIGAM )
+        if( pfh.magic == TCPDUMP_CIGAM ) {
             SWAP32( pkh.caplen );
+            SWAP32( pkh.len );
+        }
 
         n = pkh.caplen;
 
diff --git a/src/packetforge-ng.c b/src/packetforge-ng.c
index 294c160..81c382c 100644
--- a/src/packetforge-ng.c
+++ b/src/packetforge-ng.c
@@ -202,8 +202,10 @@ int capture_ask_packet( int *caplen )
             return( 1 );
         }
 
-        if( dev.pfh_in.magic == TCPDUMP_CIGAM )
+        if( dev.pfh_in.magic == TCPDUMP_CIGAM ) {
             SWAP32( pkh.caplen );
+            SWAP32( pkh.len );
+        }
 
         tv.tv_sec  = pkh.tv_sec;
         tv.tv_usec = pkh.tv_usec;
diff --git a/src/tkiptun-ng.c b/src/tkiptun-ng.c
index 5c0e80b..2a34753 100644
--- a/src/tkiptun-ng.c
+++ b/src/tkiptun-ng.c
@@ -1097,8 +1097,10 @@ int capture_ask_packet( int *caplen, int just_grab )
                 return( 1 );
             }
 
-            if( dev.pfh_in.magic == TCPDUMP_CIGAM )
+            if( dev.pfh_in.magic == TCPDUMP_CIGAM ) {
                 SWAP32( pkh.caplen );
+                SWAP32( pkh.len );
+            }
 
             tv.tv_sec  = pkh.tv_sec;
             tv.tv_usec = pkh.tv_usec;
-- 
1.8.0

